### ISSUE TITLE
deno fetch fails when destination path can be file also directory

### ISSUE THREAD

**Issue by keroxp** [denoland/deno#2790]
```
deno fetch https://dev.jspm.io/react-dom
Download https://dev.jspm.io/react-dom
Download https://dev.jspm.io/npm:react-dom@16.9.0/index.dew.js
Download https://dev.jspm.io/npm:react-dom@16.9.0/cjs/react-dom.development.dew.js
Download https://dev.jspm.io/npm:react@16?dew
Download https://dev.jspm.io/npm:object-assign@4?dew
Download https://dev.jspm.io/npm:prop-types@15/checkPropTypes?dew
Download https://dev.jspm.io/npm:scheduler@0.15?dew
Download https://dev.jspm.io/npm:scheduler@0.15/tracing?dew
Download https://dev.jspm.io/npm:object-assign@4.1.1/index.dew.js
Download https://dev.jspm.io/npm:prop-types@15.7.2/checkPropTypes.dew.js
Download https://dev.jspm.io/npm:prop-types@15.7.2/lib/ReactPropTypesSecret.dew.js
Download https://dev.jspm.io/npm:scheduler@0.15.0/tracing.dew.js
Download https://dev.jspm.io/npm:scheduler@0.15.0/cjs/scheduler-tracing.development.dew.js
thread 'tokio-runtime-worker-6' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 21, kind: Other, message: "Is a directory" }', src/libcore/result.rs:999:5
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
thread 'main' panicked at 'internal error: entered unreachable code', ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.22/src/runtime/threadpool/mod.rs:296:26
thread 'tokio-runtime-worker-0' panicked at 'unexpected state while aborting task: Complete', ../../third_party/rust_crates/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.15/src/task/mod.rs:216:21
```

I don't know precisely what is going on, but something goes wrong..

Content of `~/Library/Caches/deno/deps/https/dev.jspm.io/npm:scheduler@0.15.headers.json` is:
```
{"mime_type":"application/javascript"}
```

but `~/Library/Caches/deno/deps/https/dev.jspm.io/npm:scheduler@0.15` is DIRECTORY

This may occur when there are two url modules like these:
- https://dev.jspm.io/npm:scheduler@0.15 <- application/javascript
- https://dev.jspm.io/npm:scheduler@0.15/tracing <-application/javascript

**Comment by kevinkassimo** [comment]
Panics here: https://github.com/denoland/deno/blob/4faab6a74b0f583bba5ebcc877c6ea1407d360e3/cli/file_fetcher.rs#L380-L386

https://dev.jspm.io/npm:scheduler@0.15?dew does not do any redirect. Instead it just serves a file. Thus in the cache file `npm:scheduler@0.15` is created, prohibiting any attempt to create a folder of the same name. (thus exploding at `npm:scheduler@0.15/tracing`)

This feels more like the `index.js/index.html` scenario to me. Unfortunately jspm chooses direct serving instead of triggering redirect. I am worried that this might become a common problem at some point.

Maybe we should provide a fallback such that when we have a filename vs directory name collision, we move the file (thus making room for the directory) to another name which could still be accessed (e.g. `npm:scheduler@0.15` file renamed to `npm:schedular@0.15.index`) and try look up this fallback filename every time when we cannot find a file. Whether to do such compromise is honestly debatable, but I am kind of anxious that this problem might be common enough at some point (since there is no required mapping between URL path and file path, and that there isn't really the required concept of `directory` or `file` when serving content)

Another side problem I notice here is that we do not put query params into cache paths (e.g. trailing `?dew` is gone when saving local file). Servers obviously can choose to serve different content under different query. @bartlomieju do you think we should encode query as part of the file path? (I believe '?' is an invalid char in filenames on windows?)

**Comment by bartlomieju** [comment]
@kevinkassimo nice catch! There is already an issue for that, ref #2763

**Comment by keroxp** [comment]
@kevinkassimo Thank you. I think this is quite difficult problem. One simple solution is to store file flatly with full url sha hash (without subdirectories). But this idea seems to have some inconvenience.

**Comment by bartlomieju** [comment]
Fixed in #4030, pending `v0.34.0` release
